Run ILC after ComputeResolvedFilesToPublishList

sbomer opened the pull request:
Fixes https://github.com/dotnet/runtime/issues/108909

Needs more testing, but seems to work with a small repro so far.

Fixes https://github.com/dotnet/runtime/issues/118298 (removes CopyNativeBinary logic)

sbomer committed:
Run ILC after ComputeResolvedFilesToPublishList

MichalStrehovsky commented:
Did you consider using the [scheme that ILLink/R2R/SingleFile use right now](https://github.com/dotnet/sdk/blob/4bc2a9b5a6b32d275a487f8175cf3bdfbdc35b8b/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Publish.targets#L446-L471)?

It feels like it would make the sequencing most "standard". We'd probably need to wait for the new SDK to flow to this repo but we just did an upgrade and we could do another one.

agocke commented on src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Publish.targets:
```diff
@@ -13,24 +13,50 @@
+    </PropertyGroup>
+
+    <ItemGroup Condition="'$(CopyOutputSymbolsToPublishDirectory)' == 'true'">
+      <_symbolRecursivePath Include="$(NativeOutputPath)$(TargetName)$(_symbolExt)/**/*" />
```
I think it may be possible to simplify this -- I'll have to refresh my MSbuild knowledge to remember if it's easy to just track everything in a subdir.

sbomer committed:
Fix symbol folder RelativePath

am11 commented on src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Publish.targets:
```diff
@@ -13,24 +13,50 @@
+
+      <!-- Add native symbol file if it exists. On Mac, the symbol path may be a folder-->
+      <ResolvedFileToPublish Include="$(_symbolSourcePath)"
+                             Condition="'$(_symbolIsFile)' == 'true' and Exists('$(_symbolSourcePath)')">
+        <RelativePath>$(TargetName)$(_symbolExt)</RelativePath>
+        <CopyToPublishDirectory>PreserveNewest</CopyToPublishDirectory>
+      </ResolvedFileToPublish>
+
+      <!-- Add folder otherwise -->
+      <ResolvedFileToPublish Include="@(_symbolRecursivePath)"
+                             Condition="'$(_symbolIsFile)' == 'false' and Exists('$(_symbolSourcePath)')">
+        <RelativePath>$(TargetName)$(_symbolExt)/%(RecursiveDir)%(Filename)%(Extension)</RelativePath>
+        <CopyToPublishDirectory>PreserveNewest</CopyToPublishDirectory>
```
nit: limit the scope of condition

```suggestion
                             Condition="Exists('$(_symbolSourcePath)')">
        <!-- If the symbol is a dSYM bundle, it's a folder not a file. -->
        <RelativePath Condition="'$(NativeSymbolExt)' == '.dSYM'">$(TargetName)$(_symbolExt)</RelativePath>
        <RelativePath Condition="'$(NativeSymbolExt)' != '.dSYM'">$(TargetName)$(_symbolExt)</RelativePath>
        <CopyToPublishDirectory>PreserveNewest</CopyToPublishDirectory>
```

am11 commented on src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Publish.targets:
```diff
@@ -13,24 +13,50 @@
+      <_symbolExt Condition="'$(OS)' != 'Windows_NT'">$(NativeBinaryExt)$(NativeSymbolExt)</_symbolExt>
+      <_symbolSourcePath>$(NativeOutputPath)$(TargetName)$(_symbolExt)</_symbolSourcePath>
+      <_symbolTargetPath>$(PublishDir)\$(TargetName)$(_symbolExt)</_symbolTargetPath>
+      <!-- If the symbol is a dSYM bundle, it's a folder not a file. -->
+      <_symbolIsFile Condition="'$(NativeSymbolExt)' == '.dSYM'">false</_symbolIsFile>
+      <_symbolIsFile Condition="'$(_symbolIsFile)' == ''">true</_symbolIsFile>
```
```suggestion
```

sbomer committed:
Limit scope of condition

MichalStrehovsky commented on src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Publish.targets:
```diff
@@ -92,40 +115,4 @@
 
   </Target>
 
-  <Target Name="CopyNativeBinary" AfterTargets="Publish">
```
Any opinion on leaving this as an empty target to make this less breaking? There are some hits on Github with dependencies on this target. In .NET 10, I think everyone can just switch to depending on Publish, so the fix is easy, but maybe tolerating a line here with a comment for a couple release would be fine too.

Example: https://github.com/microsoft/node-api-dotnet/blob/50061dd8778a1b186900a41a0487e7db5582693d/src/NodeApi/NodeApi.csproj#L37

MichalStrehovsky reviewed (approved):
Nice!

MichalStrehovsky commented:
(The CI failures look related)

sbomer committed:
Oops

sbomer committed:
Keep empty CopyNativeBinary target

sbomer committed:
Fix ios test

sbomer committed:
Merge remote-tracking branch 'origin/main' into ilcPublish

ivanpovazan commented:
/azp run runtime-ioslike

ivanpovazan commented:
/azp run runtime-extra-platforms

ivanpovazan commented:
> @ivanpovazan is there going to be any work required in the iOS SDK as a result of this change?

No, because we opt-out from NativeAOT Publish targets and handle them differently:
https://github.com/xamarin/xamarin-macios/blob/8d57fe69ca1d9d63c8da2c9b5672eda4ea69ab4d/dotnet/targets/Xamarin.Shared.Sdk.props#L189

ivanpovazan commented:
@sbomer ios/tvos test failures seem related, let me know if I can help with those.

sbomer committed:
Fix target order

sbomer committed:
Move to LibraryBuilder.targets

ivanpovazan commented on src/mono/msbuild/apple/build/AppleBuild.props:
```diff
@@ -36,6 +36,7 @@
         $(_ProcessRuntimeComponentsForLibraryMode);
         $(_AotCompileTargetName);
         _BuildNativeLibrary;
+        _PublishNativeBinary;
```
Remove this change, see my comment below.

ivanpovazan commented on src/mono/msbuild/common/LibraryBuilder.targets:
```diff
@@ -46,6 +46,17 @@
     </LibraryBuilderTask>
   </Target>
 
+  <Target Name="_PublishNativeBinary">
```
I don't think we should put this target here. 

`LibraryBuilder.props/targets` are intended for Mono's library mode support, and none of the MSBuild targets from this file should be needed for NativeAOT (I admit there are other hacks which are advocating against this).

If I understand correctly, you added this target so that NativeAOT library mode has the previous `CopyNativeBinary` behavior, correct? 

If that is the case, then I would:
1. Move this target into `AppleBuild.targets`
2. Remove it from `AppleBuildDependsOn` (see my comment above)
3. Add it manually as a last dependency at: https://github.com/dotnet/runtime/blob/572533b433aa1b0b5a498cec2d310b939acd3a93/src/mono/msbuild/apple/build/AppleBuild.targets#L276
4. Put a condition on it: `Condition="'$(_IsLibraryMode)' == 'true' and '$(UseNativeAOTRuntime)' == 'true'"`

This way it does not affect: Mono app or lib builds, and it does not affect NativeAOT app builds. We will eventually clean all of this up in the internal props/targets as we now have experimental CoreCLR support for Apple as well.

sbomer committed:
Merge remote-tracking branch 'origin/main' into ilcPublish

# Conflicts:
#	src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Publish.targets

sbomer committed:
Address PR feedback: move _PublishNativeBinary to AppleBuild.targets and clean up stale references

- Move _PublishNativeBinary from LibraryBuilder.targets to AppleBuild.targets,
  gated on _IsLibraryMode and UseNativeAOTRuntime, appended to
  _AppleNativeAotCompile DependsOnTargets.
- Remove _PublishNativeBinary from AppleBuildDependsOn in AppleBuild.props.
- Remove stale _CopyAotSymbols reference from AndroidBuildDependsOn.

sbomer committed:
Skip non-PostprocessAssembly files in NativeAOT managed assembly classification

Content items that happen to be managed DLLs (e.g. TestAssembly.dll
referenced with OutputItemType=Content) were incorrectly classified as
managed assemblies to compile and removed from publish output. Use the
SDK's PostprocessAssembly metadata to only classify files the SDK has
identified as runtime assemblies, and add ReferenceOutputAssembly=false
to the FileVersionInfo test so the SDK doesn't treat it as one.

sbomer commented:
/azp run runtime-android

sbomer commented:
I think this is ready to go - runtime-extra-platforms failures look unrelated (fixing some of them in https://github.com/dotnet/runtime/pull/124582). PTAL @ivanpovazan @MichalStrehovsky @agocke

agocke reviewed (approved):
lGTM, thanks

MichalStrehovsky reviewed (approved):
Thank you!

sbomer committed:
Merge remote-tracking branch 'origin/main' into ilcPublish
